csrc/scheduler/utils.cpp

```diff
@@ -1337,7 +1337,9 @@ IterDomain* projectIdToRoot(
     return nullptr;
   }
 
-  if (!tv->hasRoot()) {
+  // Note we check for `hasAllocation`, because allocation domain suggests there
```

Comment:
Does reference_id always belong to getMaybeAllocationDomain? If yes, I'd rename the function to projectAllocationIdToRoot to reflect that. If not, the `!tv->hasAllocation()` check might be problematic.

Reply:
> Does reference_id always belong to getMaybeAllocationDomain?

I don't think so. Since we use PairwiseLogicalDomainMap, I think reference_id can be on logical domain as well, when we project from C2P.

What's the issue with `hasAllocation` check?

I'm assuming `getExprsTo({reference_id})` would still lead me to IDs on root domain, whether reference id is on logical or allocation.

Reply:
> What's the issue with hasAllocation check?

You are right -- there's nothing wrong with it. I thought this check was meant for correctness. 

After reading the rest of the code, I think this check can be safely removed for simplicity. Performance is likely neutral because, when there's no root and no allocation, getExprsTo returns an empty set immediately. I tend to aggressively remove `if`s unless they are proven to be important.

Reply:
Yeah I understand as these are meant to have an early return if it's unlikely to find any expressions within the tensor domain.

I agree with you. let me try removing those.